File: src/ExpoBaiduMap.types.ts
import type { ColorValue, StyleProp, ViewStyle } from "react-native";

export type OnLoadEventPayload = {
  url: string;
};

export type ExpoBaiduMapModuleEvents = {
  onChange: (params: ChangeEventPayload) => void;
  onGetPoiCitySearchResult: (
    result: PoiSearchResult | { errorCode: number },
  ) => void;
  onGetPoiNearbySearchResult: (
    result: PoiSearchResult | { errorCode: number },
  ) => void;
  onGetPoiBoundsSearchResult: (
    result: PoiSearchResult | { errorCode: number },
  ) => void;
  onGetPoiDetailSearchResult: (
    result: PoiDetailResult | { errorCode: number },
  ) => void;
  onGetSuggestionResult: (
    result: PoiSuggestionResult | { errorCode: number },
  ) => void;
  onGetGeoCodeResult: (result: GeoCodeResult | { errorCode: number }) => void;
  onGetReGeoCodeResult: (
    result: ReGeoCodeResult | { errorCode: number },
  ) => void;
};

export type ChangeEventPayload = {
  value: string;
};

export type MapType = "standard" | "satellite" | "none";

export type UserTrackingMode = "heading" | "follow" | "followWithHeading";

export type CoordinateType = "gps" | "common" | "bd09ll";

export type LogoPosition =
  | "leftBottom"
  | "leftTop"
  | "centerBottom"
  | "centerTop"
  | "rightBottom"
  | "rightTop";

export type ExpoBaiduMapViewProps = {
  style?: StyleProp<ViewStyle>;
  active: boolean;
  mapType?: MapType;
  language?: "english" | "chinese";
  backgroundColor?: ColorValue;
  backgroundImage?: string;
  region?: CoordinateRegion;
  limitMapRegion?: CoordinateRegion;
  compassPosition?: Point;
  centerCoordinate?: Coordinate2D;
  showsUserLocation?: boolean;
  inDoorMapEnabled?: boolean;
  userTrackingMode?: UserTrackingMode;
  // Map style properties
  showsScale?: boolean;
  showsZoomControls?: boolean;
  rotationEnabled?: boolean;
  doubleClickZoomEnabled?: boolean;
  showsCompass?: boolean;
  logoPosition?: LogoPosition;
  logoOffset?: Point;
  scrollEnabled?: boolean;
  zoomEnabled?: boolean;
  tiltEnabled?: boolean;
  maxZoomLevel?: number;
  minZoomLevel?: number;
  // Overlays
  markers?: Marker[];
  polylines?: Polyline[];
  arcs?: Arc[];
  polygons?: Polygon[];
  circles?: Circle[];
};

export type Coordinate2D = {
  latitude: number;
  longitude: number;
};

export type CoordinateSpan = {
  latitudeDelta: number;
  longitudeDelta: number;
};

export type CoordinateRegion = {
  center: Coordinate2D;
  span: CoordinateSpan;
};

export type Point = {
  x: number;
  y: number;
};

export type Size = {
  width: number;
  height: number;
};

/**
 * 点标记覆盖物。
 */
export type Marker = {
  /**
   * 点标记的坐标点。
   */
  coordinate: Coordinate2D;
  /**
   * 点标记的唯一标识符。必填，因为你需要区分多个点标记。
   */
  key: string;
  /**
   * 点标记的标题。
   */
  title?: string;
  /**
   * 点标记的副标题。
   */
  subtitle?: string;
  /**
   * 点标记的图标。
   */
  icon?: string;
  /**
   * 点标记的旋转角度。
   */
  rotation?: number;
  /**
   * 点标记的透明度。
   */
  alpha?: number;
  /**
   * 点标记是否可点击。
   */
  isClickable?: boolean;
  /**
   * 点标记的Z轴顺序。
   */
  zIndex?: number;
};

/**
 * 折线覆盖物。
 */
export type Polyline = {
  /**
   * 折线的坐标点数组。
   */
  coordinates: Coordinate2D[];
  /**
   * 折线的颜色。
   */
  strokeColor: string;
  /**
   * 折线的宽度。
   */
  strokeWidth: number;
  /**
   * 折线是否可点击。
   */
  isClickable?: boolean;
  /**
   * 折线的Z轴顺序。
   */
  zIndex?: number;
};

/**
 * 弧线覆盖物。
 */
export type Arc = {
  /**
   * 弧线的坐标点数组。
   */
  coordinates: Coordinate2D[];
  /**
   * 弧线的颜色。
   */
  strokeColor: string;
  /**
   * 弧线的宽度。
   */
  lineWidth: number;
  /**
   * 弧线的Z轴顺序。
   */
  zIndex?: number;
};

/**
 * 多边形覆盖物。根据一堆坐标点生成一个多边形覆盖物。
 */
export type Polygon = {
  /**
   * 多边形的坐标点数组。
   */
  coordinates: Coordinate2D[];
  /**
   * 多边形的填充颜色。
   */
  fillColor: string;
  /**
   * 多边形的边框颜色。
   */
  strokeColor: string;
  /**
   * 多边形的边框宽度。
   */
  strokeWidth: number;
  /**
   * 多边形是否可点击。
   */
  isClickable?: boolean;
  /**
   * 多边形的Z轴顺序。
   */
  zIndex?: number;
};

/**
 * 圆覆盖物。根据一个坐标点和半径生成一个圆覆盖物。
 */
export type Circle = {
  /**
   * 圆的中心坐标点。
   */
  center: Coordinate2D;
  /**
   * 圆的半径。
   */
  radius: number;
  /**
   * 圆的填充颜色。
   */
  fillColor: string;
  /**
   * 圆的边框颜色。
   */
  strokeColor: string;
  /**
   * 圆的边框宽度。
   */
  strokeWidth: number;
  /**
   * 圆是否可点击。
   */
  isClickable?: boolean;
  /**
   * 圆的Z轴顺序。
   */
  zIndex?: number;
};


export type Bounds = {
  northEast: Coordinate2D;
  southWest: Coordinate2D;
};

export type PoiSearchScope = "basic" | "detail";

export type PoiCitySearchOptions = {
  city: string;
  keyword: string;
  pageIndex?: number;
  pageSize?: number;
  scope?: PoiSearchScope;
  tag?: string;
  cityLimit?: boolean;
  returnAddress?: boolean;
};

export type PoiNearbySearchOptions = {
  keyword: string;
  location: Coordinate2D;
  radius?: number;
  pageIndex?: number;
  pageSize?: number;
  scope?: PoiSearchScope;
  tag?: string;
  radiusLimit?: boolean;
  returnAddress?: boolean;
};

export type PoiBoundsSearchOptions = {
  keyword: string;
  bounds: Bounds;
  pageIndex?: number;
  pageSize?: number;
  scope?: PoiSearchScope;
  tag?: string;
  returnAddress?: boolean;
};

export type PoiDetailSearchOptions = {
  uid: string;
};

export type PoiSuggestionOptions = {
  keyword: string;
  city?: string;
  cityLimit?: boolean;
  location?: Coordinate2D;
  hotWord?: boolean;
};

export type PoiSearchPagination = {
  pageIndex: number;
  pageSize: number;
  totalCount: number;
  totalPage: number;
};

export type PoiInfo = {
  uid: string;
  name: string;
  location?: Coordinate2D;
  address?: string;
  province?: string;
  city?: string;
  area?: string;
  phone?: string;
  tag?: string;
};

export type PoiSearchResult = {
  pagination: PoiSearchPagination;
  pois: PoiInfo[];
};

export type PoiDetailInfo = {
  uid: string;
  name: string;
  location?: Coordinate2D;
  address?: string;
  province?: string;
  city?: string;
  area?: string;
  phone?: string;
  tag?: string;
};

export type PoiDetailResult = {
  poi: PoiDetailInfo;
};

export type PoiSuggestion = {
  uid?: string;
  key?: string;
  city?: string;
  district?: string;
  location?: Coordinate2D;
};

export type PoiSuggestionResult = {
  suggestions: PoiSuggestion[];
};

export type GeoCoderOptions = {
  address: string;
  city?: string;
  output?: "json" | "xml";
  retCoordType?: CoordinateType;
};

export type ReGeoCoderOptions = {
  location: Coordinate2D;
  radius?: number;
  tags?: string[];
  extensionsRoad?: boolean;
  pageSize?: number;
  pageNum?: number;
  isLatestAdmin?: boolean;
};

export type GeoCodeResult = {
  location: Coordinate2D;
  precise: number;
  confidence: number;
  level: string;
};

export type ReGeoCodeAddressComponent = {
  country: string;
  countryCode: string;
  countryCodeISO: string;
  countryCodeISO2: string;
  province: string;
  city: string;
  cityLevel: number;
  district: string;
  town: string;
  townCode: string;
  street: string;
  streetNumber: string;
  direction: string;
  distance: string;
  adcode: string;
  adcodeFull: string;
  provinceCode: string;
  cityCode: string;
  districtCode: string;
};

export type ReGeoCodeResult = {
  formattedAddress: string;
  addressComponent: ReGeoCodeAddressComponent;
  business: string;
  pois: PoiInfo[];
  roads: any[];
  roadIntersections: any[];
  aois: any[];
};

File: ios/ExpoBaiduMap.types.swift
//
//  ExpoBaiduMap.types.swift
//  ExpoBaiduMap
//
//  Created by Aron on 2025/12/8.
//

import Foundation
import ExpoModulesCore
import BaiduMapAPI_Base
import BaiduMapAPI_Map
import BaiduMapAPI_Search

struct Coordinate2D: Record {
    init() { }
    init(latitude: Double = 0.0, longitude: Double = 0.0) {
        self.latitude = latitude
        self.longitude = longitude
    }
    
    @Field
    var latitude = 0.0
    @Field
    var longitude = 0.0
    
    func toCLCoordinate() -> CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }
}

struct CoordinateSpan: Record {
    init() { }
    init(latitudeDelta: Double = 0.0, longitudeDelta: Double = 0.0) {
        self.latitudeDelta = latitudeDelta
        self.longitudeDelta = longitudeDelta
    }
    
    @Field
    var latitudeDelta = 0.0
    @Field
    var longitudeDelta = 0.0
}

struct CoordinateRegion: Record {
    @Field
    var center: Coordinate2D = .init()
    @Field
    var span: CoordinateSpan = .init()
    
    func toBMKRegion() -> BMKCoordinateRegion {
        return BMKCoordinateRegion(center: CLLocationCoordinate2D(latitude: center.latitude,
                                                                  longitude: center.longitude),
                                   span: BMKCoordinateSpan(latitudeDelta: span.latitudeDelta,
                                                           longitudeDelta: span.longitudeDelta))
    }
}

struct Point: Record {
    init() { }
    init(x: Double = 0.0, y: Double = 0.0) {
        self.x = x
        self.y = y
    }
    
    @Field
    var x = 0.0
    @Field
    var y = 0.0
}

struct Size: Record {
    init() { }
    init(width: Double = 0.0, height: Double = 0.0) {
        self.width = width
        self.height = height
    }
    
    @Field
    var width = 0.0
    @Field
    var height = 0.0
}

struct Bounds: Record {
    init() { }
    
    @Field
    var northEast: Coordinate2D = .init()
    
    @Field
    var southWest: Coordinate2D = .init()
}

struct Polygon: Record {
    init() { }
    init(coordinates: [Coordinate2D] = [], count: Int = 0) {
        self.coordinates = coordinates
        self.count = count
    }
    
    @Field
    var coordinates: [Coordinate2D] = []
    @Field
    var count = 0
}

final class Circle: Record {
    init() { }
    init(center: Coordinate2D = .init(), radius: Double = 0.0) {
        self.center = center
        self.radius = radius
    }
    
    @Field
    var center: Coordinate2D = .init()
    @Field
    var radius = 0.0
    @Field
    var fillColor: UIColor?
    @Field
    var strokeColor: UIColor?
    
    @Field
    var lineWidth: CGFloat?
}

final class TextMarker: Record {
    
    @Field
    var center: Coordinate2D = .init()
    
    @Field
    var text: String?
    
    @Field
    var textColor: UIColor?
    
    @Field
    var textSize: CGFloat?
    
    @Field
    var backgroundColor: UIColor?
}

struct PoiCitySearchOptions: Record {
    init() { }
    
    @Field
    var city: String = ""
    
    @Field
    var keyword: String = ""
    
    @Field
    var pageIndex: Int?
    
    @Field
    var pageSize: Int?
    
    @Field
    var scope: String?
    
    @Field
    var tag: String?
    
    @Field
    var cityLimit: Bool?
    
    @Field
    var returnAddress: Bool?
}

struct PoiNearbySearchOptions: Record {
    init() { }
    
    @Field
    var keywords: [String] = []
    
    @Field
    var location: Coordinate2D = .init()
    
    @Field
    var radius: Int?
    
    @Field
    var pageIndex: Int?
    
    @Field
    var pageSize: Int?
    
    @Field
    var scope: String?
    
    @Field
    var tag: String?
    
    @Field
    var radiusLimit: Bool?
    
    @Field
    var tags: [String]?
}

struct PoiBoundsSearchOptions: Record {
    init() { }
    
    @Field
    var keywords: [String] = []
    
    @Field
    var bounds: Bounds = .init()
    
    @Field
    var pageIndex: Int?
    
    @Field
    var pageSize: Int?
    
    @Field
    var scope: String?
    
    @Field
    var tags: [String]?
    
    @Field
    var returnAddress: Bool?
}

struct PoiDetailSearchOptions: Record {
    init() { }
    
    @Field
    var uids: [String] = []
}

struct PoiSuggestionOptions: Record {
    init() { }
    
    @Field
    var keyword: String = ""
    
    @Field
    var city: String?
    
    @Field
    var cityLimit: Bool?
    
    @Field
    var location: Coordinate2D?
}

struct PoiSearchPagination: Record {
    init() { }
    
    @Field
    var pageIndex: Int = 0
    
    @Field
    var pageSize: Int = 0
    
    @Field
    var totalCount: Int = 0
    
    @Field
    var totalPage: Int = 0
}

struct PoiInfoRecord: Record {
    init() { }
    
    @Field
    var uid: String = ""
    
    @Field
    var name: String = ""
    
    @Field
    var location: Coordinate2D?
    
    @Field
    var address: String?
    
    @Field
    var province: String?
    
    @Field
    var city: String?
    
    @Field
    var area: String?
    
    @Field
    var phone: String?
    
    @Field
    var tag: String?
}

struct PoiSearchResultRecord: Record {
    init() { }
    
    @Field
    var pagination: PoiSearchPagination = .init()
    
    @Field
    var pois: [PoiInfoRecord] = []
}

struct PoiDetailInfoRecord: Record {
    init() { }
    
    @Field
    var uid: String = ""
    
    @Field
    var name: String = ""
    
    @Field
    var location: Coordinate2D?
    
    @Field
    var address: String?
    
    @Field
    var province: String?
    
    @Field
    var city: String?
    
    @Field
    var area: String?
    
    @Field
    var phone: String?
    
    @Field
    var tag: String?
}

struct PoiDetailResultRecord: Record {
    init() { }
    
    @Field
    var poiList: [PoiDetailInfoRecord] = []
}

struct PoiSuggestionRecord: Record {
    init() { }
    
    @Field
    var uid: String?
    
    @Field
    var key: String?
    
    @Field
    var city: String?
    
    @Field
    var district: String?
    
    @Field
    var location: Coordinate2D?
}

struct PoiSuggestionResultRecord: Record {
    init() { }
    
    @Field
    var suggestions: [PoiSuggestionRecord] = []
}
